Ray implementation of maple inference pipeline

kaylahardie opened the pull request:
To run the ray implementation of the maple inference pipeline:

- Use the "environment_maple.yml" file to create a conda environment with ray: 
run `conda env create -f environment_maple.yml` and then `conda activate maple_py310_ray`
- Create the directory structures by running `mpl_workflow_create_dir_struct.py`
- Add a sample input image to the data/input_img_local directory, I used the sample image here https://drive.google.com/file/d/1YwQiPc7Ow-oSyEHuCBD97RxRbnzJ4_dW/view?usp=drive_link  
- run `python3 maple_workflow.py --gpus_per_core=0` for running the pipeline on a cpu, I haven't tried running it on more than one cpu. I've only tried running it on one image. When running it on two images locally it ran out of memory
- The results from the pipeline should be in the data/ray_shapefiles dir, you can use compare_shapefile_features.py to compare the features in two shapefiles or you can use 'ogrinfo -so -al <path to shapefile>' on the command line to examine a shapefile.

kaylahardie committed:
Create ray conda environment

kaylahardie committed:
Format file

kaylahardie committed:
Load geotiffs into ray dataset

kaylahardie committed:
Calculate Water Mask

kaylahardie committed:
Apply water mask and tile image

kaylahardie committed:
Starting to rayify inference but there's a bug

kaylahardie committed:
Fix inference ray bug

kaylahardie committed:
Fixed creating image tiles -> needed to use deepcopy, also made changes to how the dataclasses represented the dict

kaylahardie committed:
Fix inference -> Ray throws type errors when we set a column value to be a List

kaylahardie committed:
Remove old inference code

kaylahardie committed:
Stitching image tiles back together using Ray

kaylahardie committed:
Close gdal files after opening them fix and also added code to write shapefiles

kaylahardie committed:
Remove unnecessary code from stitch shapefile

kaylahardie committed:
fix gdal virtual file system issue - the gdal vfs only works in a single process so instead of having the virtual file in the ray dataset, adapted the code to store the image bytes in the dataset and create the gdal virtual file locally when needed

kaylahardie committed:
Break logic up into different files and fix imports

kaylahardie committed:
Rename files, add compare_unordered_shapefiles for testing purposes

kaylahardie committed:
Add back original maple_workflow file

kaylahardie committed:
fix file name indexing for shapefile

kaylahardie committed:
Add documentation for compare_unordered_shapefiles

kaylahardie committed:
Adding more documentation

kaylahardie committed:
More documentation

kaylahardie committed:
More comments

kaylahardie committed:
Making it possible to read and write to gcs buckets when running ray_maple_workflow.py

kaylahardie committed:
add application default credentials directory as a flag so it isn't hardcoded, this is needed for service account impersonation if running the code on your local computer and want it to be able to access gcs buckets

kaylahardie commented:
The most recent [commit](https://github.com/PermafrostDiscoveryGateway/MAPLE_v3/pull/19/commits/f43a4b6b5edf9608920c0714810658f3c26ae09d) makes it possible to run the ray version of the maple pipeline using [pdg's google cloud storage bucket](https://console.cloud.google.com/storage/browser/pdg-storage-default/workflows_optimization/maple_ray_pipeline/data/ray_output_shapefiles;tab=objects?project=pdg-project-406720&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false) instead of your local filesystem.

Here are the instructions for **locally running** (ie. running on your local computer) the ray version of the maple pipeline with access to gcs storage buckets:
To give the ray program that's running on your local computer access to the gcs storage bucket, it's best practice to use [service account impersonation](https://cloud.google.com/docs/authentication/use-service-account-impersonation):
1. make sure your email that has access to the [pdg-project](https://console.cloud.google.com/welcome?project=pdg-project-406720) also has the [roles/iam.serviceAccountTokenCreator](https://cloud.google.com/iam/docs/service-account-permissions#token-creator-role). 
2. run ```gcloud auth application-default login --impersonate-service-account=pdg-sa-01@pdg-project-406720.iam.gserviceaccount.com --scopes="https://www.googleapis.com/auth/cloud-platform"``` It will ask you to sign in and then you will get access to an application default credentials directory. The terminal should print something like ```Credentials saved to file: [/usr/local/google/home/kaylahardie/.config/gcloud/application_default_credentials.json]```
3. Run ```python3 ray_maple_workflow.py --gpus_per_core=0 --root_dir="gs://pdg-storage-default/workflows_optimization/maple_ray_pipeline" --adc_dir="/usr/local/google/home/kaylahardie/.config/gcloud/application_default_credentials.json"``` but replace the directory that you got from step 2 into the --adc_dir flag.
4. when the program finishes the output shapefiles will be stored in the directory: [gs://pdg-storage-default/workflows_optimization/maple_ray_pipeline/data/ray_output_shapefiles/](https://console.cloud.google.com/storage/browser/pdg-storage-default/workflows_optimization/maple_ray_pipeline/data/ray_output_shapefiles?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&project=pdg-project-406720)

Right now it just overwrites the old outputted files, we can add some versioning to the output file names to avoid overwriting the old files if desired. It takes as input, the data in the gs://pdg-storage-default/workflows_optimization/maple_ray_pipeline/data/input_img/ directory

kaylahardie committed:
Each run is now outputted under it's own datetime directory

kaylahardie commented:
Most recent [commit](https://github.com/PermafrostDiscoveryGateway/MAPLE_v3/pull/19/commits/5cdf5c01094e576f702283bd277bb419a9fd0774) makes it so each run doesn't overwrite the data from previous runs. Each run now creates a date directory in the format "%Y-%m-%d_%H-%M-%S" in the [gs://pdg-storage-default/workflows_optimization/maple_ray_pipeline/data/ray_output_shapefiles/](https://console.cloud.google.com/storage/browser/pdg-storage-default/workflows_optimization/maple_ray_pipeline/data/ray_output_shapefiles?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&project=pdg-project-406720) directory and in the date directory the output shapefiles for that run are stored

kaylahardie committed:
small fixes (trying to limit the changes on the original maple_workflow.py). Changed the input img directory back to be the input_img_local (what it was originally)

gugibugy commented on compare_shapefile_features.py:
```diff
@@ -0,0 +1,60 @@
+def feature_to_hashable(feature):
+    """Converts a feature into a hashable representation."""
+    geometry_repr = shape(feature['geometry']).wkt
+    attributes = tuple(sorted(feature['properties'].items()))
```
is there a benefit to having this sorted and converted into a tuple or can it stay as an unsorted list?

kaylahardie replied:
lists aren't hashable so I needed to convert it to a tuple to be able to sort it in the set. I sorted the list to make the set comparison possible

gugibugy commented on gdal_virtual_file_system.py:
```diff
@@ -0,0 +1,27 @@
+import os
+from osgeo import gdal
+
+class GDALVirtualFileSystem:
```
General idea: Maybe instead of having the create and close methods you can do the following. Move all the logic of create into the init function and move the close function into the destructor. That way users just have to create the object and when it gets garbage collected the gdal stuff gets cleaned up. Might provide for a cleaner interface!

kaylahardie replied:
hey I thought about something similar when I first implemented this but got stuck on how python handles destructors/clean up. I found a better solution though, using with statements: https://stackoverflow.com/questions/865115/how-do-i-correctly-clean-up-a-python-object. What do you think?

gugibugy replied:
I think that would provide a nice interface. Specially since it has a lot of parallels to how most people interact with file objects (which is through "with" statements).

gugibugy commented on mpl_config.py:
```diff
@@ -48,6 +52,22 @@ def __init__(
         self.TEMP_W_IMG_DIR = self.ROOT_DIR + r"/data/water_mask/temp"
         self.OUTPUT_IMAGE_DIR = self.ROOT_DIR + r"/data/output_img"
         self.WORKER_ROOT = self.ROOT_DIR + r"/data"
+        self.MODEL_DIR = self.ROOT_DIR + r"/local_dir/datasets/logs"
```
This looks like weird directory to have the model in...

kaylahardie replied:
I agree, that's the location of it according to the maple code, maybe that could be changed in a future iteration of the code

gugibugy commented on ray_image_preprocessing.py:
```diff
@@ -0,0 +1,136 @@
+"""
+
+
+import os
+from pathlib import Path
+import shutil
+from typing import Dict, Any
+
+import numpy as np
+from osgeo import gdal
+from skimage import filters, io
+from skimage.morphology import disk
+
+from mpl_config import MPL_Config
+import gdal_virtual_file_system as gdal_vfs
```
nit: Usually it's nice to have the raw "imports" at the top and then following by the "from ... import ..."

gugibugy replied:
This applies to all files in the PR. It would also be great if we could make the effort to alphabetize the imports!

kaylahardie replied:
hey - I was following the python style guide of:

Imports should be grouped in the following order:

Standard library imports.
Related third party imports.
Local application/library specific imports.

https://peps.python.org/pep-0008/#imports

I will make sure they're alphabetical within that ordering

gugibugy commented on ray_image_preprocessing.py:
```diff
@@ -0,0 +1,136 @@
+    worker_water_subroot = os.path.join(worker_water_root, image_name)
+    temp_water_subroot = os.path.join(temp_water_root, image_name)
+    # Prepare to make directories to create the files
+    try:
+        shutil.rmtree(worker_water_subroot)
+    except:
+        print("directory deletion failed")
+        pass
+
+    try:
+        shutil.rmtree(temp_water_subroot)
+    except:
+        print("directory deletion failed")
+        pass
+
+    # check local storage for temporary storage
+    Path(worker_water_subroot).mkdir(parents=True, exist_ok=True)
+    Path(temp_water_subroot).mkdir(parents=True, exist_ok=True)
```
if we are creating the directories using "exist_ok=true" then we should be able to remove the directory deletion logic above. Or is the deletion there to ensure that we don't have any artifacts present from previous runs?

tcnichol replied:
If I remember I think that was the reason - things left over from previous runs.

kaylahardie replied:
Instead of doing it this way, I think it makes more sense to keep data from previous runs but have them dated so you know which data came from which runs. That might be helpful to know how the data produced from the different runs is changing over time. I'm not going to address that in this PR since it's out of scope but I'd recommend it for future work

kaylahardie replied:
I did the dating approach in the ray_write_shapefiles.py in case you're interested

gugibugy commented on ray_image_preprocessing.py:
```diff
@@ -0,0 +1,136 @@
+        print("gdal Translate failed with", gdal.GetLastErrorMsg())
+        pass
+
+    image = io.imread(
+        output_tif_8b_file
+    )  # image[rows, columns, dimensions]-> image[:,:,3] is near Infrared
```
it would be great if we didn't have to read the output_tif_8b_file back right after writing it! Could this be avoided by writing the output_tif_8b_file to a "virtual" file?

kaylahardie replied:
the reason why they chosen io.imread is because it creates a numpy array. I altered the code to use gdal but I'm not using the GdalVirtualFilePath class that I created since gdal.translate handles writing to the virtual file......I'm wondering if this gdal version of the code makes the code less readable, wdyt?

kaylahardie replied:
I'm going to resolve this

gugibugy commented on ray_image_preprocessing.py:
```diff
@@ -0,0 +1,136 @@
+
+    bilat_img = filters.rank.median(nir, disk(value))
+
+    gtif = gdal.Open(virtual_file_path)
+    geotransform = gtif.GetGeoTransform()
+    sourceSR = gtif.GetProjection()
```
it would make it more readable if this was moved closer to where geotransform and sourceSR are referenced.

kaylahardie replied:
I agree but is it nice to grab the fields and then see the del/file close closer to opening the file?

kaylahardie replied:
I ended up using with statements since that seems to be the pythonic approach https://gis.stackexchange.com/questions/80366/why-close-a-dataset-in-gdal-python

gugibugy reviewed (commented):
First pass, will take a look at the rest on Monday.

gugibugy commented on ray_infer_tiles.py:
```diff
@@ -0,0 +1,109 @@
+        # Used to identify a specific predictor when mulitple predictors are
+        # created to run inference in parallel. The counter is also used to
+        # know which GPU to use when multiple are available.
+        self.process_counter = 1  # TODO need to fix this process_counter
```
This should be replaced by a call to ray.get_gpu_ids which will return a list of available GPUs. For now we could simply choose a value at random (if this become more complex we can re-adress this).

kaylahardie replied:
wouldn't we want to keep track of which ids have been used? if we just pick a random id we could randomly pick the same gpu right?

gugibugy commented on ray_infer_tiles.py:
```diff
@@ -0,0 +1,109 @@
+from mpl_config import MPL_Config, PolygonConfig
+
+
+@dataclass
+class ShapefileResult:
+    polygons: np.array
+    class_id: int
+
+
+@dataclass
+class ShapefileResults:
+    shapefile_results: List[ShapefileResult]
```
These data classes are also declared  in the ray_tile_and_stitch_util.py. I would say they should be declared in one place and imported by other modules that reference it.

kaylahardie replied:
done

gugibugy commented on ray_tile_and_stitch_util.py:
```diff
@@ -0,0 +1,341 @@
+    xsize = input_image_gtif.RasterXSize
+
+    # Close the file.
+    input_image_gtif = None
```
I've seen a couple of instances where this pattern is used to remove the reference to the image so that the file gets closed. I think that using the built in "del" keyword might do a better job at conveying what we are trying to accomplish with this line. If you opt to go with this, make sure you apply it to the other places where this pattern is used.

kaylahardie replied:
done

gugibugy commented on ray_tile_and_stitch_util.py:
```diff
@@ -0,0 +1,341 @@
+                cols = block_size
+            else:
+                cols = xsize - j
+            # get block out of the whole raster
+            # todo check the array values is similar as ReadAsArray()
+            band_1_array = final_array_4[i: i + rows, j: j + cols]
+            band_2_array = final_array_2[i: i + rows, j: j + cols]
+            band_3_array = final_array_3[i: i + rows, j: j + cols]
+
+            # filter out black image
+            if (
+                band_3_array[0, 0] == 0
+                and band_3_array[0, -1] == 0
+                and band_3_array[-1, 0] == 0
+                and band_3_array[-1, -1] == 0
+            ):
+                continue
+
+            # stack three bands into one array
+            img = np.stack((band_1_array, band_2_array, band_3_array), axis=2)
+            cv2.normalize(img, img, 0, 255, cv2.NORM_MINMAX)
```
I think that all this code could be simplified if we created the image stack above when we are reading the bands from the input geotif. If you did that, you could apply the mask dot product to all channels in one operations and getting the image would become something like 
```
img = final_array[i:i+rows, j:j+cols, :]
```

kaylahardie replied:
tried doing that and ran into weird errors with numpy so I'm going to not implement this suggestion, it's also tangential to the ray migration. Thanks for the idea though :)

gugibugy commented on ray_tile_and_stitch_util.py:
```diff
@@ -0,0 +1,341 @@
+        polygon_dict[k] = [poly_count, poly_count + v]
+        poly_count += v
+
+    first_row = group.head(1)
+    image_data_from_arbitrary_row = first_row["image_metadata"][0]
+    size_i, size_j = image_data_from_arbitrary_row.len_y_list, image_data_from_arbitrary_row.len_x_list
```
Do we choose an arbitrary row because they all have the same value and it doesn't matter where we read it from?

kaylahardie replied:
yes because we grouped by image, all the rows (each corresponding to a specific tile) should have the same image metadata

gugibugy commented on ray_tile_and_stitch_util.py:
```diff
@@ -0,0 +1,341 @@
+
+    close_list = list()
+    print("Total number of polygons: ", len(centroid_list))
+    tile_blocksize = 4
+
+    for id_i in range(0, size_i, 3):
```
It would be great to be able to have some comments around where this magical blocksize of 4 comes from and why the loops in the section below have steps of 3.

kaylahardie replied:
yeah I agree, @amalshehan can you remind us why those values were chosen? I know you mentioned it before but our slack history disappeared

gugibugy commented on ray_tile_and_stitch_util.py:
```diff
@@ -0,0 +1,341 @@
+    close_list = [list(x) for x in close_list]
+
+    # --------------- looking for connected components in a graph ---------------
+    def connected_components(lists):
+        neighbors = defaultdict(set)
+        seen = set()
+        for each in lists:
+            for item in each:
+                neighbors[item].update(each)
+
+        def component(node, neighbors=neighbors, seen=seen, see=seen.add):
+            nodes = set([node])
+            next_node = nodes.pop
+            while nodes:
+                node = next_node()
+                see(node)
+                nodes |= neighbors[node] - seen
+                yield node
+
+        for node in neighbors:
+            if node not in seen:
+                yield sorted(component(node))
```
there is likely a library we can use to do this computation, instead of implementing it ourselves and having to manage it.

kaylahardie replied:
ack though I feel like that can be done in a separate pr

gugibugy commented on ray_tile_and_stitch_util.py:
```diff
@@ -0,0 +1,341 @@
+    return new_rows
+
+
+def stitch_shapefile(group: pd.DataFrame):
+    """
+    Create a shapefile for each image.
+    Note that normally ray rows are dictionaries but this is a group because we've called
+    grouped by. When ray does a groupby and the rows can't be represented in numpy arrays
+    it uses pandas dataframe.
+
+    Parameters
+    ----------
+    group: a pandas dataframe that has all of the shapefile information for each tile in the
+            image.
+    """
```
This documentation is extremely helpful! It would be even better if we could add some information around the output as well since it's not entirely clear why we return the "first_row" of the group.

kaylahardie replied:
ack done :)

gugibugy commented on ray_write_shapefiles.py:
```diff
@@ -0,0 +1,169 @@
+            vfs = gdal_vfs.GDALVirtualFileSystem(
+                file_path, file_bytes)
+            virtual_file_path = vfs.create_virtual_file()
+            print("here is the virtual file path: ", virtual_file_path)
```
nit: is this useful in prod?

kaylahardie replied:
removed

gugibugy commented on ray_write_shapefiles.py:
```diff
@@ -0,0 +1,169 @@
+            # Get the coordinate system information
+            wkt = self.get_coordinate_system_info(geotiff_path, geotiff_bytes)
+
+            print(wkt)
```
nit: is this useful in prod?

kaylahardie replied:
removed

gugibugy commented on ray_write_shapefiles.py:
```diff
@@ -0,0 +1,169 @@
+        except Exception as e:
+            print(f"Error when trying to write the prj file: {e}")
+
+    def write_shapefile_helper(self, row: Dict[str, Any], w):
```
nit: type hint for "w" would be very helpful!

gugibugy replied:
Also, this seems to be a private function of the class. Although you can't enforce private functions, it is common practice to prefix private function names with "__" to indicate to others that they shouldn't be calling this function from outside the class.

kaylahardie replied:
good call!

kaylahardie replied:
honestly there are a lot of private functions that should have the __ prefix

gugibugy commented on ray_write_shapefiles.py:
```diff
@@ -0,0 +1,169 @@
+        if self.config.GCP_FILESYSTEM is None:
+            delete_and_create_dir(self.config.RAY_SHAPEFILES)
+
+    def get_coordinate_system_info(self, file_path: str, file_bytes: bytes):
```
seems to be a private function, see my comment below!

kaylahardie replied:
done

gugibugy commented on ray_write_shapefiles.py:
```diff
@@ -0,0 +1,169 @@
+            writer = shapefile.Writer(f"{shapefile_output_dir_for_image}.shp")
+            self.write_shapefile_helper(row, writer)
+            writer.close()
+
+
+
+
```
nit: remove unnecessary white space.

kaylahardie replied:
done

gugibugy commented on ray_maple_workflow.py:
```diff
@@ -0,0 +1,130 @@
+    )
+
+    print("0. Load geotiffs into ray dataset")
+    dataset = create_geotiff_images_dataset(config).map(add_image_name)
```
how would this work when a directory of images is used to run the pipeline?

kaylahardie replied:
it creates one row for each image in the dir

gugibugy commented on ray_maple_workflow.py:
```diff
@@ -0,0 +1,130 @@
+    print("0. Load geotiffs into ray dataset")
+    dataset = create_geotiff_images_dataset(config).map(add_image_name)
+    print("Ray dataset schema:", dataset.schema())
+    print("1. Start calculating watermask")
+    dataset_with_water_mask = dataset.map(fn=ray_image_preprocessing.cal_water_mask,
+                                          fn_kwargs={"config": config})
+    print("Dataset schema with water mask: ", dataset_with_water_mask.schema())
+    print("2. Start tiling image")
+    image_tiles_dataset = dataset_with_water_mask.flat_map(
+        fn=ray_tile_and_stitch_util.tile_image, fn_kwargs={"config": config})
+    image_tiles_dataset = image_tiles_dataset.drop_columns(["mask"])
+    print("Dataset schema with image tiles: ", image_tiles_dataset.schema())
+    print("3. Start inferencing")
+    inferenced_dataset = image_tiles_dataset.map(
+        fn=ray_inference.MaskRCNNPredictor, fn_constructor_kwargs={"config": config}, concurrency=2)
+    print("Dataset schema with inferenced tiles: ", inferenced_dataset.schema())
+    print("4. Start stitching")
+    data_per_image = inferenced_dataset.groupby(
+        "image_name").map_groups(ray_tile_and_stitch_util.stitch_shapefile)
+    print("Dataset schema where each row is an image (result of a group by tile): ",
+          data_per_image.schema())
+    print("5. Write shapefiles")
+    shapefiles_dataset = data_per_image.map(
+        fn=ray_write_shapefiles.WriteShapefiles, fn_constructor_kwargs={"config": config}, concurrency=2)
+    print("Done writing shapefiles", shapefiles_dataset.schema())
```
In most of the map calls, concurrency is not being set. I think it is important to make a note that the concurrency parameter should be tuned based on the resources available in the raycluster in order to make the best use of the compute resources available. If no concurrency is set, it just runs a single worker for it which won't provide much improvement.

kaylahardie replied:
added an arg for concurrency

kaylahardie committed:
addressing
gugibugy's review comments

gugibugy commented on ray_maple_workflow.py:
```diff
@@ -0,0 +1,134 @@
+    )
+
+    print("Ray pipeline starting")
+    ray_dataset = create_geotiff_images_dataset(config).map(
```
I would argue that this level of nested chaining isn't very readable. You can still break this up into separate intermediate variables. Just make sure you chain the output together and not define new datasets while do so. Take a look at the ray_maple_workflow.py in the scaling-ray-inference branch for an example of what I mean.

kaylahardie replied:
I didn't like the nested chaining either, it now looks way better :D

gugibugy commented on ray_write_shapefiles.py:
```diff
@@ -0,0 +1,155 @@
+        self.config = config
+        self.current_timestamp_str = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+        if self.config.GCP_FILESYSTEM is None:
+            delete_and_create_dir(self.config.RAY_OUTPUT_SHAPEFILES_DIR)
```
This logic makes it so that concurrently running actors might accidentally delete each other's directories (ran into this my self). We should try to come up with a less destructive way of doing this or one that plays nicer with concurrency.

kaylahardie replied:
yeah your'e right - this is problematic - thanks for point it out :)
I moved the create dir to main since we only need to check if the dir exists and create it once (no need to check for each actor). I don't delete the output dir beacuse the output files have unique dates for each run. I think it's nicer to have the output from multiple runs to be able to compare. 

There is still work to be done for creating gcp dirs if they don't exist. I left a comment and a todo for that. The directory management could be cleaned up in future work

kaylahardie committed:
addressing more comments, fixing delete ray shapefile dir

tcnichol commented:
I deleted then created the virtual environment over again since it looked like it created it with the wrong python version. To fix this I made a small commit and added the python version in the yml file.

Trying to run it, I'm getting this error:

```
usage: ray_maple_workflow.py [-h] --host HOST --dataset_id DATASET_ID --key
                             KEY [--image <command>] [--root_dir ROOT_DIR]
                             [--weight_file WEIGHT_FILE]
                             [--gpus_per_core GPUS_PER_CORE]
ray_maple_workflow.py: error: the following arguments are required: --host, --dataset_id, --key
```

It's complaining about not having these values, but if we're running things with google cloud, shouldn't those be irrelevant?

tcnichol commented:
This one looks good to me, I've been able to run everything.

tcnichol reviewed (approved):
I'm going to mark this approved. It looks like all comments were addressed, runs fine, environment works.

gugibugy commented on ray_maple_workflow.py:
```diff
@@ -107,26 +117,48 @@ def add_image_name(row: Dict[str, Any]) -> Dict[str, Any]:
+            create_directory_if_not_exists(config.RAY_OUTPUT_SHAPEFILES_DIR)
+    shapefiles_dataset = data_per_image.map(
+        fn=ray_write_shapefiles.WriteShapefiles, fn_constructor_kwargs={"config": config}, concurrency=concurrency)
+    print("MAPLE Ray pipeline finished, done writing shapefiles", shapefiles_dataset.schema())
```
You'll actually want to call materialize here instead of schema. Materialize actually runs the pipeline for all the rows in the dataset, while my understanding is that schema only materializes a single row in order to get the schema of the data set.

kaylahardie replied:
How'd you find out the schema only materializes a single row? I updated the code, I'm just curious

gugibugy replied:
When I was prototyping with running on more than 1 image, if you did .schema() it would only produce results for a single row

kaylahardie committed:
Updated readme

kaylahardie commented:
I updated the README as well a little bit. The [Building Scalable ML Pipelines Using Ray](https://docs.google.com/document/d/14kru_GPsakDJFX-On1Db_NLClEFvB6oLFMQFFkNQqsA/edit?usp=sharing) is also updated. Going to merge now

kaylahardie committed:
fix merge conflict with weight file

kaylahardie committed:
updating readme again

kaylahardie committed:
Merge branch 'main' into try-ray-inference
